project13/ViewController.swift

```diff
@@ -4,16 +4,191 @@
+        
+        let beginImage = CIImage(image: currentImage)
+        currentFilter.setValue(beginImage, forKey: kCIInputImageKey)
+        applyProcessing()
+        stopLoadingAnimation()
+        
+    }
+    
+    @IBAction func changeFilter(_ sender: UIButton) {
+        let alertController = UIAlertController(title: "Choose filter", message: nil, preferredStyle: .actionSheet)
+        CIFilterName.allCases.forEach { filter in
+                  let action = UIAlertAction(title: filter.stringName, style: .default, handler: { _ in
+                      self.set(filter: filter)
+                  })
+                  alertController.addAction(action)
+              }
+        if let popoverController = alertController.popoverPresentationController {
+            popoverController.sourceView = sender
+            popoverController.sourceRect = sender.bounds
+        }
+        present(alertController, animated: true)
+    }
+    
+    @IBAction func save(_ sender: UIButton) {
+        guard let image = imageView.image else {
+            let alertController = UIAlertController(title: "Error", message: "Please selct an image", preferredStyle: .actionSheet)
+            alertController.addAction(UIAlertAction(title: "Ok", style: .destructive))
+            present(alertController, animated: true)
+            return
+        }
+        UIImageWriteToSavedPhotosAlbum(image, self, #selector(image(_:didFinishSavingWithError: contextInfo:)),nil)
+    }
+    
+    @IBAction func intensityChanger(_ sender: Any) {
+        applyProcessing()
+    }
+    
+    private func applyProcessing() {
+        
+       guard let currentImage = currentImage else { return }
+        guard let currentFilter = currentFilter else { return }
+        let inputKeys = currentFilter.inputKeys
+        
+        if inputKeys.contains(kCIInputIntensityKey){
+            currentFilter.setValue(intensity.value, forKey: kCIInputIntensityKey)
+        }
+        if inputKeys.contains(kCIInputRadiusKey){
+            currentFilter.setValue(intensity.value * 200, forKey: kCIInputRadiusKey)
+        }
+        if inputKeys.contains(kCIInputScaleKey) {
+            currentFilter.setValue(intensity.value * 10, forKey: kCIInputScaleKey)
+        }
+        if inputKeys.contains(kCIInputCenterKey) {
+            currentFilter.setValue(CIVector(x: currentImage.size.width / 2, y: currentImage.size.height / 2), forKey: kCIInputCenterKey)
+        }
+        
+        guard let outputImage = currentFilter.outputImage else { return }
+        
+        if let cgImage = context.createCGImage(outputImage, from: outputImage.extent) {
+            let processedImage = UIImage(cgImage: cgImage)
+            imageView.image = processedImage
+        }
+    }
+    
+    private func set(filter: CIFilterName) {
+        currentFilter = CIFilter.make(filter)
+        applyProcessing()
+    }
+}
+enum CIFilterName :String, CaseIterable {
```

Comment:
Nitpick: we usually have the colon right after the type name, without spaces. On the other hand, there should be a space between the colon and superclass name/protocol conformances.

`enum CIFilterName: String, CaseIterable {`

Reply:
This looks more like a typo actually :P

Reply:
And it was haha! 
Thank you!